othello/moderator/runners.py

```diff
@@ -0,0 +1,167 @@
+import os
+import select
+import signal
+import subprocess
+import time
+
+import psutil
+
+from django.conf import settings
+
+from ..sandboxing import get_sandbox_args
+from .moderator import Player
+from .utils import ServerError, UserError, capture_generator_value
+
+LEGACY_MOVES = {
+    (
+        x + 11
+        if x < 8
+        else x + 13
+        if x < 16
+        else x + 15
+        if x < 24
+        else x + 17
+        if x < 32
+        else x + 19
+        if x < 40
+        else x + 21
+        if x < 48
+        else x + 23
+        if x < 56
+        else x + 25
+    ): x
+    for x in range(64)
+}
+
+
+def legacy_board_convert(board):
+    legacy_board = ""
+    for i in range(100):
+        if i in LEGACY_MOVES:
+            legacy_board += board[LEGACY_MOVES[i]]
+        else:
+            legacy_board += "?"
+    return legacy_board.replace(Player.BLACK.value, "@")
+
+
+class PlayerRunner:
+    def __init__(self, submission, driver):
+        if not os.path.isfile(submission.code.path):
+            raise OSError("file not found")
+        self.path = submission.code.path
+        self.is_legacy = submission.is_legacy
+        self.process = None
+        self.driver = driver
+
+    def __enter__(self):
+        self.start()
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.stop()
+
+    def start(self):
+        cmd_args = ["python3", "-u", self.driver, self.path]
+        if settings.DEBUG:
+            cmd_args = get_sandbox_args(
+                cmd_args, whitelist=[os.path.dirname(self.path)], readonly=[self.path]
+            )
+
+        self.process = subprocess.Popen(
+            cmd_args,
+            stdin=subprocess.PIPE,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            bufsize=0,
+            cwd=os.path.dirname(self.path),
+            preexec_fn=os.setpgrp,
+        )
+
+    def stop(self):
+        if self.process is not None:
+            children = psutil.Process(self.process.pid).children(recursive=True)
+            try:
+                os.killpg(self.process.pid, signal.SIGKILL)
+            except ProcessLookupError:
+                self.process.kill()
+            for child in children:
+                try:
+                    child.kill()
+                except psutil.NoSuchProcess:
+                    pass
+            self.process = None
+
+    @capture_generator_value
+    def get_move(self, board, player, time_limit, last_move):
+        if self.process.poll():
+            print(self.process.communicate())
+            return -1, ServerError.PROCESS_EXITED
+
+        if self.is_legacy:
+            board = legacy_board_convert(board)
+            player = "@" if player == Player.BLACK.value else "o"
```

Comment:
You know what, I'm seeing this "legacy translation" code so much that I think it's worth having functions to translate the player characters both ways. Perhaps `Player.to_legacy()` and `@classmethod Player.from_legacy()`?

Reply:
In regards to legacy player conversion, this is the only occurrence where I switch 'x' to '@'. 
I guess the part in `legacy_board_convert` where I do `.replace('x', '@')` can count, but that wouldn't benefit from the method you are proposing.

Reply:
I'm just not fond of sticking constants in the main protocol implementation.